torch._scaled_mm with MXFP8

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147548

# summary

Add blockwise MXFP8 support to `torch._scaled_mm` on CUDA capability 10.0 and higher devices.  If the scales for A and B are of dtype `torch.float8_e8m0fnu`, we dispatch to the blockwise kernel from cuBLAS.

This is a skeleton PR where we test basic functionality (numerics of various simple matrices, as well as one end to end quantization + gemm).


- Scales are flipped based on transpose_result 
- Handles boundary conditions 

Note that MXFP4 is not added in this PR - we can tackle that in a future PR.

This PR was created by taking https://github.com/pytorch/pytorch/pull/145562, switching e8m0 to in-core dtype, removing fp4 for now, and adding test cases.

# test plan

```
pytest test/test_matmul_cuda.py -k blockwise_mxfp8 -s
```



cc @ptrblck @msaroufim @eqy

vkuzo committed:
Update

[ghstack-poisoned]

vkuzo committed:
Update

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -1514,7 +1518,12 @@ void scaled_gemm(
-  {
+  if (mat1_scale_dtype == kFloat8_e8m0fnu && mat2_scale_dtype == kFloat8_e8m0fnu) {
+    // scales are swapped in the cuBLAS kernel, swap them here to hide this from the user
+    // TODO(future): figure out what is going on here
```
need to fix this before land

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg commented on test/test_matmul_cuda.py:
```diff
@@ -366,6 +368,74 @@ def to_fp8_saturated(
+    n_col_blocks = ceil_div(cols, 4)
+
+    # Pad out and view as tiles of (128, 4)
+    padded = torch.nn.functional.pad(input_matrix.view(torch.uint8), (0, -cols % 4, 0, -rows % 128))
```
Pad errors on fp8 going to update

drisspg committed:
Update

[ghstack-poisoned]

drisspg commented on test/test_matmul_cuda.py:
```diff
@@ -366,6 +368,79 @@ def to_fp8_saturated(
+    padded_cols = n_col_blocks * 4
+
+    padded = input_matrix
+    # Ideally we would use torch.nn.pad but it doesn't support float8_e8m0fnu for now
```
Side steps the problem as well it is slightly faster:

```Shell
Running benchmarks on: cuda

Benchmarking matrix size: 128x4
Version 1 (zeros): 2200.115 mss
Version 2 (pad):   2197.597 ms
Speedup: 1.00x
Outputs match: True
Memory usage v1: 0.01 MB
Memory usage v2: 0.01 MB

Benchmarking matrix size: 256x8
Version 1 (zeros): 3904.562 mss
Version 2 (pad):   4025.708 ms
Speedup: 0.97x
Outputs match: True
Memory usage v1: 0.04 MB
Memory usage v2: 0.05 MB

Benchmarking matrix size: 130x6
Version 1 (zeros): 2139.283 mss
Version 2 (pad):   4148.867 ms
Speedup: 0.52x
Outputs match: True
Memory usage v1: 0.03 MB
Memory usage v2: 0.04 MB

Benchmarking matrix size: 1024x32
Version 1 (zeros): 4458.569 mss
Version 2 (pad):   4725.722 ms
Speedup: 0.94x
Outputs match: True
Memory usage v1: 0.63 MB
Memory usage v2: 0.75 MB

Benchmarking matrix size: 2048x64
Version 1 (zeros): 5008.937 mss
Version 2 (pad):   5605.401 ms
Speedup: 0.89x
Outputs match: True
Memory usage v1: 2.62 MB
Memory usage v2: 3.00 MB
```

drisspg replied:
zeros: https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/drisspg_8790d5a5-c29c-40d5-bcc2-50dce9f1f2e3_zeros.json

pad: https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/drisspg_b406f046-5b19-4c6f-9574-90d861e26061_pad.json

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg commented:
@pytorchbot merge

wdvr commented:
@pytorchmergebot revert -m 'failing internal build - discussed with author' -c ghfirst

cc @vkuzo  @drisspg

wdvr commented:
@pytorchmergebot revert -m 'failing internal build - discussed with author' -c ghfirst

wdvr commented:
@pytorchmergebot revert -m 'failing internal build - discussed with author' -c ghfirst

drisspg committed:
Update

[ghstack-poisoned]

drisspg commented:
@pytorchbot merge

wdvr commented:
@pytorchmergebot revert -m 'failing internal build - similar to previous, see below' -c ghfirst


aten/src/ATen/native/cuda/Blas.cpp:103:46: error: no member named 'nullopt' in namespace 'c10'; did you mean 'std::nullopt'?
  103 |       const std::optional<Tensor>& scale_a = c10::nullopt,
      |                                              ^~~~~~~~~~~~
      |                                              std::nullopt

wdvr commented:
@pytorchmergebot revert -m 'failing internal build - similar to previous, see below' -c ghfirst

aten/src/ATen/native/cuda/Blas.cpp:103:46: error: no member named 'nullopt' in namespace 'c10'; did you mean 'std::nullopt'?
103 | const std::optional& scale_a = c10::nullopt,
| ^~~~~~~~~~~~
| std::nullopt

cc @drisspg @vkuzo

jeffdaily commented:
@drisspg would you send me a message before landing this again so I can verify it on MI300 since we don't have MI300 available pre-merge?

drisspg committed:
Update on "torch._scaled_mm with MXFP8"


# summary

Add blockwise MXFP8 support to `torch._scaled_mm` on CUDA capability 10.0 and higher devices.  If the scales for A and B are of dtype `torch.float8_e8m0fnu`, we dispatch to the blockwise kernel from cuBLAS.

This is a skeleton PR where we test basic functionality (numerics of various simple matrices, as well as one end to end quantization + gemm).


- Scales are flipped based on transpose_result 
- Handles boundary conditions 

Note that MXFP4 is not added in this PR - we can tackle that in a future PR.

This PR was created by taking https://github.com/pytorch/pytorch/pull/145562, switching e8m0 to in-core dtype, removing fp4 for now, and adding test cases.

# test plan

```
pytest test/test_matmul_cuda.py -k blockwise_mxfp8 -s
```



cc ptrblck msaroufim eqy

[ghstack-poisoned]

drisspg committed:
Update on "torch._scaled_mm with MXFP8"


# summary

Add blockwise MXFP8 support to `torch._scaled_mm` on CUDA capability 10.0 and higher devices.  If the scales for A and B are of dtype `torch.float8_e8m0fnu`, we dispatch to the blockwise kernel from cuBLAS.

This is a skeleton PR where we test basic functionality (numerics of various simple matrices, as well as one end to end quantization + gemm).


- Scales are flipped based on transpose_result 
- Handles boundary conditions 

Note that MXFP4 is not added in this PR - we can tackle that in a future PR.

This PR was created by taking https://github.com/pytorch/pytorch/pull/145562, switching e8m0 to in-core dtype, removing fp4 for now, and adding test cases.

# test plan

```
pytest test/test_matmul_cuda.py -k blockwise_mxfp8 -s
```



cc ptrblck msaroufim eqy

[ghstack-poisoned]

drisspg commented:
@jeffdaily can you try again I think I have removed all references to c10

jeffdaily commented:
@drisspg 

Suggested diff if you could apply it.  test_matmul_cuda.py passes after this diff is applied.  

Before this PR scales were not part of the blas config struct which is why for hipblaslt rowwise we had been hard-coding the swap of the A and B scales.  Now you're doing the swapping based on if it's transposed, which is how it should be, so the diff needed to remove that hard-coded swap but preserve the use of the custom matmul desc attr for hipblaslt *_VEC_EXT.

In test_matmul_cuda.py there were some rocm skips on UTs that verified error messages. 
 They worked fine on rocm so I removed the skips.  I changed the skip message for the carveout test because rocm _does_ support rowwise, but not carveout.  

The important part of the diff is the CUDABlas.cpp.  Take or leave the test_matmul_cuda.py changes.

```diff
diff --git a/aten/src/ATen/cuda/CUDABlas.cpp b/aten/src/ATen/cuda/CUDABlas.cpp
index ca7b23064e6..5ac7c246756 100644
--- a/aten/src/ATen/cuda/CUDABlas.cpp
+++ b/aten/src/ATen/cuda/CUDABlas.cpp
@@ -1542,19 +1542,19 @@ void scaled_gemm(
   CuBlasLtMatmulDescriptor computeDesc(computeType, scaleType);
   computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_TRANSA, _cublasOpFromChar(transa));
   computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_TRANSB, _cublasOpFromChar(transb));
+  cublasLtMatmulDescAttributes_t matmulDescA = CUBLASLT_MATMUL_DESC_A_SCALE_POINTER;
+  cublasLtMatmulDescAttributes_t matmulDescB = CUBLASLT_MATMUL_DESC_B_SCALE_POINTER;
 #if defined(USE_ROCM) && defined(HIPBLASLT_VEC_EXT)
   if (use_rowwise) {
-    // swapped
-    computeDesc.setAttribute(HIPBLASLT_MATMUL_DESC_A_SCALE_POINTER_VEC_EXT, mat2_scale_ptr);
-    computeDesc.setAttribute(HIPBLASLT_MATMUL_DESC_B_SCALE_POINTER_VEC_EXT, mat1_scale_ptr);
+    matmulDescA = HIPBLASLT_MATMUL_DESC_A_SCALE_POINTER_VEC_EXT;
+    matmulDescB = HIPBLASLT_MATMUL_DESC_B_SCALE_POINTER_VEC_EXT;
   }
-  else
 #else
   // rowwise isn't supported using cublaslt or older hipblaslt
   TORCH_INTERNAL_ASSERT(use_rowwise == false, "rowwise scaled_gemm not supported with blaslt");
 #endif
-  computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_A_SCALE_POINTER, mat1_scale_ptr);
-  computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_B_SCALE_POINTER, mat2_scale_ptr);
+  computeDesc.setAttribute(matmulDescA, mat1_scale_ptr);
+  computeDesc.setAttribute(matmulDescB, mat2_scale_ptr);
   if (result_scale_ptr != nullptr) {
     computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_D_SCALE_POINTER, result_scale_ptr);
   }
diff --git a/test/test_matmul_cuda.py b/test/test_matmul_cuda.py
index 5a02365abd0..950104a351e 100644
--- a/test/test_matmul_cuda.py
+++ b/test/test_matmul_cuda.py
@@ -708,7 +708,6 @@ class TestFP8MatmulCuda(TestCase):
         )

     @unittest.skipIf(not PLATFORM_SUPPORTS_FP8 or IS_WINDOWS, f8_msg)
-    @skipIfRocm()
     def test_float8_error_messages(self, device) -> None:
         M, K, N = (1024, 512, 2048)
         fill_value = 0.5
@@ -847,7 +846,7 @@ class TestFP8MatmulCuda(TestCase):
         self.assertEqual(out_dtype, out_fp8.dtype)
         self.assertEqual(out_fp32, out_fp8.to(torch.float))

-    @unittest.skipIf(TEST_WITH_ROCM, "ROCm doesn't support row-wise scaling")
+    @unittest.skipIf(TEST_WITH_ROCM, "ROCm doesn't support sm carveout")
     @unittest.skipIf(IS_WINDOWS, "Windows doesn't support row-wise scaling")
     @unittest.skipIf(not PLATFORM_SUPPORTS_FP8, f8_msg)
     @unittest.skipIf(not SM90OrLater, "sm89 kernel isn't opted into carveout yet")
@@ -1092,7 +1091,6 @@ class TestFP8MatmulCuda(TestCase):
             assert sqnr.item() > 22.0

     @unittest.skipIf(not PLATFORM_SUPPORTS_FP8 or IS_WINDOWS, f8_msg)
-    @skipIfRocm()
     def test_blockwise_mxfloat8_error_messages(self, device) -> None:
         M, K, N = (1024, 512, 2048)
         BLOCK_SIZE_K = 32
```

ngimel commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -95,11 +96,33 @@ c10::MaybeOwned<Tensor> inline prepare_matrix_for_cublas(const Tensor& tensor, b
     bool transpose_result = false, transpose_mat1 = false, transpose_mat2 = false;
     result = prepare_matrix_for_cublas(c, transpose_result);
     mata = prepare_matrix_for_cublas(transpose_result ? mat2 : mat1, transpose_mat1, transpose_result);
     matb = prepare_matrix_for_cublas(transpose_result ? mat1 : mat2, transpose_mat2, transpose_result);
```
Can you make it so that mat1/mat2/variable names using mat1/mat2 are not used below this point? It's been super confusing before, and now with more logic the mixture of a/b and 1/2 that don't necessarily refer to the same thing is even harder to understand.

drisspg replied:
ahhh sorry you mentioned this to me, I will update

eqy commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -1584,6 +1585,15 @@ void scaled_gemm(
   }
 
+  if (mat1_scale_dtype == kFloat8_e8m0fnu && mat2_scale_dtype == kFloat8_e8m0fnu) {
+#if CUDA_VERSION >= 12080
```
Does it make sense to gate a ROCM version similarly here CC @jeffdaily 
Otherwise is it just an #if 1 due to different version interpretation on ROCM?

jeffdaily replied:
For better or worse we hipify CUDA_VERSION to TORCH_HIP_VERSION which effectively is our ROCm distro version.  Frankly the comparison is bogus.  TODO for later.  Here it would be 603 >= 12080 which is false so ROCm will fall back to the TORCH_CHECK(false) here.

drisspg replied:
So we should have some "MIN_MX_SUPPORTED_ARCH" (name pending) defined above that we compare against?

petrex replied:
```suggestion
#if !defined(USE_ROCM) && CUDA_VERSION >= 12080
```
We could have a seperate version check for ROCm

jeffdaily replied:
It would make ROCm devs' lives easier if pytorch didn't do any CUDA version checks either in C++ or python and instead had feature macros or attrs.  What the cuda version checks are doing is guarding against new features/APIs.  Anywhere in pytorch that checks a cuda version we have to scrutinize and put either a defined(USE_ROCM) [enable feature for rocm] or !defined(USE_ROCM) [disable].

drisspg committed:
Update on "torch._scaled_mm with MXFP8"


# summary

Add blockwise MXFP8 support to `torch._scaled_mm` on CUDA capability 10.0 and higher devices.  If the scales for A and B are of dtype `torch.float8_e8m0fnu`, we dispatch to the blockwise kernel from cuBLAS.

This is a skeleton PR where we test basic functionality (numerics of various simple matrices, as well as one end to end quantization + gemm).


- Scales are flipped based on transpose_result 
- Handles boundary conditions 

Note that MXFP4 is not added in this PR - we can tackle that in a future PR.

This PR was created by taking https://github.com/pytorch/pytorch/pull/145562, switching e8m0 to in-core dtype, removing fp4 for now, and adding test cases.

# test plan

```
pytest test/test_matmul_cuda.py -k blockwise_mxfp8 -s
```



cc ptrblck msaroufim eqy

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg commented:
@pytorchbot merge
